internal/yml/promote/merge_test.go

```diff
@@ -0,0 +1,174 @@
+package promote_test
+
+import (
+	"bytes"
+	"testing"
+
+	"github.com/nestoca/joy/internal/yml"
+	"github.com/nestoca/joy/internal/yml/promote"
+	"github.com/stretchr/testify/require"
+	"gopkg.in/yaml.v3"
+)
+
+func TestYmlMerge(t *testing.T) {
+	cases := []struct {
+		Name     string
+		Src      string
+		Dst      string
+		Joy      string
+		Proposal string
+	}{
+		{
+			Name:     "conflicting types",
+			Src:      "{key: 1}",
+			Dst:      "{key: hello}",
+			Joy:      "{key: 1}",
+			Proposal: "{key: 1}",
+		},
+		{
+			Name:     "conflicting types locked dst",
+			Src:      "{key: 1}",
+			Dst:      "{key: !lock hello}",
+			Joy:      "{key: !lock hello}",
+			Proposal: "{key: !lock hello}",
+		},
+		{
+			Name:     "conflicting types inner locked dst",
+			Src:      "{key: 1}",
+			Dst:      "{key: [!lock hello]}",
+			Joy:      "{key: 1}",
+			Proposal: "{key: 1}",
+		},
+		{
+			Name:     "conflicting types locked src",
+			Src:      "{key: !lock 1}",
+			Dst:      "{key: hello}",
+			Joy:      "{key: hello}",
+			Proposal: "{key: hello}",
+		},
+		{
+			Name:     "seq",
+			Src:      "[1, 2, 3]",
+			Dst:      "[]",
+			Joy:      "[1, 2, 3]",
+			Proposal: "[1, 2, 3]",
+		},
+		{
+			Name:     "locked seq",
+			Src:      "!lock [1, 2, 3]",
+			Dst:      "[]",
+			Joy:      "!lock [1, 2, 3]",
+			Proposal: "!lock [!!int TODO, !!int TODO, !!int TODO]",
```

Comment:
What is going on here? Or would we throw an error?

Reply:
Here the source is a locked sequence and the destination has no items.

So the result is a sequence with the same length as the source but with all items marked as TODO. The !!int tag is there simply because we set the value to TODO but the node was an int.

Alternatively it could be better to set the result to: !!seq TODO

But currently this version is better than writing our locked values into the destination.